File: c/examples/tcp_sink.c
/* SPDX-FileCopyrightText: 2026 Alexander R. Croft */
/* SPDX-License-Identifier: MIT */

#include "ratatouille.h"

#include <stdio.h>

int main(void) {
    rat_tcp_sink_config_t sink_cfg = {0};
    rat_config_t cfg = {0};
    rat_tcp_sink_t *sink;
    rat_logger_t *log;
    rat_tcp_sink_stats_t sink_stats;

    sink_cfg.endpoint = "tcp://127.0.0.1:9000";
    sink = rat_tcp_sink_create(&sink_cfg);
    if (!sink) {
        fprintf(stderr, "failed to create TCP sink\n");
        return 1;
    }

    cfg.filter = "api*";
    cfg.format = RAT_FORMAT_NDJSON;
    cfg.sink = rat_tcp_sink_callback;
    cfg.sink_userdata = sink;
    cfg.source.app = "example";
    cfg.source.where = "c";
    cfg.source.instance = "local";

    log = rat_logger_create(&cfg);
    if (!log) {
        fprintf(stderr, "failed to create logger\n");
        rat_tcp_sink_destroy(sink);
        return 1;
    }

    rat_log(log, "api", "sent over tcp");
    rat_logf(log, "api", "value=%d", 42);

    sink_stats = rat_tcp_sink_stats(sink);
    fprintf(
        stderr,
        "sent=%llu failed=%llu\n",
        (unsigned long long)sink_stats.sent,
        (unsigned long long)sink_stats.failed
    );

    rat_logger_destroy(log);
    rat_tcp_sink_destroy(sink);
    return 0;
}

File: rust/examples/tcp_sink.rs
// SPDX-FileCopyrightText: 2026 Alexander R. Croft
// SPDX-License-Identifier: MIT

use ratatouille::{Format, Logger, LoggerConfig, SourceIdentity, TcpSink, TcpSinkConfig};

fn main() -> std::io::Result<()> {
    let sink = TcpSink::new(TcpSinkConfig {
        endpoint: "tcp://127.0.0.1:9000".into(),
    })?;

    let mut config = LoggerConfig::default();
    config.filter = Some("api*".into());
    config.format = Format::Ndjson;
    config.source = SourceIdentity {
        app: Some("example".into()),
        r#where: Some("rust".into()),
        instance: Some("local".into()),
    };

    let mut log = Logger::with_sink(config, sink);
    let _ = log.log("api", "sent over tcp");
    let _ = log.log("api", "value=42");

    let stats = log.sink_mut().stats();
    eprintln!("sent={} failed={}", stats.sent, stats.failed);
    Ok(())
}